README.md
# ArcOvh

Use Arc with the [OVH Cloudstorage](https://www.ovh.co.uk/public-cloud/storage/)


## Installation

The package can be installed by adding `arc_ovh` to your
list of dependencies in `mix.exs`:

```elixir
def deps do
  [{:arc_ovh, "~> 0.2"}]
end
```


## Configuration

  - Configure the Arc Storage adapter
```elixir
config :arc,
  storage: Arc.Storage.Ovh.Cloudstorage,
  client: ArcOvh.Client.Cloudstorage,
  container: "default",
  default_tempurl_ttl: (30 * 24 * 60 * 60), # 30 days default time to live for signed urls.
  version_timeout: (60 * 3 * 1000) # 3 minutes
```

  - Configure the OVH client
```elixir
config :arc_ovh, ArcOvh.Client.Cloudstorage,
    adapter: Openstex.Adapters.Ovh,
    ovh: [
      application_key: System.get_env("APPLICATION_KEY"),
      application_secret: System.get_env("APPLICATION_SECRET"),
      consumer_key: System.get_env("CONSUMER_KEY")
    ],
    keystone: [
      tenant_id: System.get_env("TENANT_ID"), # mandatory, corresponds to an ovh project id or ovh servicename
      user_id: System.get_env("USER_ID"), # optional, if absent a user will be created using the ovh api.
      endpoint: "https://auth.cloud.ovh.net/v2.0"
    ],
    swift: [
      account_temp_url_key1: System.get_env("TEMP_URL_KEY1"), # defaults to :nil if absent
      account_temp_url_key2: System.get_env("TEMP_URL_KEY2"), # defaults to :nil if absent
      region: :nil # specify a region if you like. Check your OVH cloud.
    ],
    hackney: [
      timeout: 20000,
      recv_timeout: 40000
    ]

config :httpipe,
  adapter: HTTPipe.Adapters.Hackney
```


## Example usage

  - Create Definition (see [arc](https://github.com/stavro/arc) for more information)

```elixir
defmodule DummyDefinitionThumbnail do
    use Arc.Definition
    @versions [:original, :thumbnail]

    def transform(:thumbnail, _) do
      {"convert", "-strip -thumbnail 100x100^ -gravity center -extent 100x100 -format jpg", :jpg}
    end
    # Whitelist file extensions:
    def validate({file, _}) do
      file_extension = file.file_name |> Path.extname() |> String.downcase()
      file_extension in ~w(.jpg .jpeg .gif .png)
    end
    def __storage(), do: Application.get_env(:arc, :storage, Arc.Storage.Ovh.Cloudstorage)

    def storage_dir(_version, {_file, scope}) do
      (scope != :nil && Map.has_key?(scope, :id)) &&
       "#{scope.id}" ||
       ""
    end
    def storage_dir(_version, _) do
       ""
    end

    def filename(version, {file, _scope}) do
      "#{Path.basename(file.file_name) |> Path.rootname()}_#{version}"
    end
end
```

   - Store an image and generate a url for the
   `:original` version and `:thumbnail` version

```elixir
defp image(), do: Path.join(__DIR__, "./assets/image.jpg") |> Path.expand()

{:ok, object_name} = DummyDefinitionThumbnail.store({image(), %{id: 1}})
original_url = DummyDefinitionThumbnail.url({image(), %{id: 1}}, :original)
thumbnail_url = DummyDefinitionThumbnail.url({image(), %{id: 1}}, :thumbnail)
```

   - Store an image and generate a signed url for the
   `:original` version and `:thumbnail` version

```elixir
defp image(), do: Path.join(__DIR__, "./assets/image.jpg") |> Path.expand()

{:ok, object_name} = DummyDefinitionThumbnail.store({image(), %{id: 1}})
original_signed_url = DummyDefinitionThumbnail.url({image(), %{id: 1}}, :original, signed: :true)
thumbnail_signed_url = DummyDefinitionThumbnail.url({image(), %{id: 1}}, :thumbnail, signed: :true)
```

   - Delete an image for all versions.

```elixir
{:ok, object_name} = DummyDefinitionThumbnail.store({image(), %{id: 1}})
:ok = DummyDefinitionThumbnail.delete({object_name, %{id: 1}})
```


## Tests

- Tests are run against a container named `arc_ovh_test_container` on an ovh
openstack server.

- Create a `test.exs` file in `config.exs` and setup a [Configuration](## Configuration)

- `mix test`

- The image used in the tests was taken in bundoran Ireland by me a few years back.

<p align="center">
  <img src="https://github.com/stephenmoloney/arc_ovh/blob/master/test/fixtures/bundoran.jpg?raw=true" width="600">
</p>



## License

[MIT Licence](LICENSE)


CHANGELOG.md
# Changelog

## v0.2.0

[changes]
- Update dependencies for elixir 1.5+

## v0.1.5

[changes]
- bump openstex_adapters_ovh dependency

## v0.1.4

[changes]
- bump dependencies
- Remove `Plug.MIME.path/1` in favour of `MIME.path/1`

## v0.1.3

[changes]
- bump dependencies

## v0.1.2

[changes]
- bump version of `:openstex_adapters_ovh` to `0.3.5`
- Add `:openstex_adapters_ovh` to list of applications to avoid
mix release warnings.


## v0.1.1

[bugfix]
- Delete url when the scope is not passed. Before, failed due to absent scope.

[documentation]
- remove use of pseudofolder from documentation as it is
completely optional.


## v0.1.0

- Initial commit
- Consists of:
    - `ArcOvh.Client.Cloudstorage`, an OVH client
    - `ArcOvh.Application`, an Application to supervise the OVH client
    - `Arc.Storage.Ovh.Cloudstorage`, an adapter for storage for [arc](https://github.com/stavro/arc)
- Basic docs in readme

config/test.exs
use Mix.Config

config :arc_ovh, ArcOvh.Client.Cloudstorage,
    adapter: Openstex.Adapters.Ovh,
    ovh: [
      application_key: System.get_env("APPLICATION_KEY_TEST"),
      application_secret: System.get_env("APPLICATION_SECRET_TEST"),
      consumer_key: System.get_env("CONSUMER_KEY_TEST")
    ],
    keystone: [
      tenant_id: System.get_env("TENANT_ID_TEST"), # mandatory, corresponds to an ovh project id or ovh servicename
      user_id: System.get_env("USER_ID_TEST"), # optional, if absent a user will be created using the ovh api.
      endpoint: "https://auth.cloud.ovh.net/v2.0"
    ],
    swift: [
      account_temp_url_key1: System.get_env("TEMP_URL_KEY1"), # defaults to :nil if absent
      account_temp_url_key2: System.get_env("TEMP_URL_KEY2"), # defaults to :nil if absent
      region: :nil
    ],
    hackney: [
      timeout: 20000,
      recv_timeout: 40000
    ]

config :httpipe,
  adapter: HTTPipe.Adapters.Hackney

config :arc,
  storage: Arc.Storage.Ovh.Cloudstorage,
  client: ArcOvh.Client.Cloudstorage,
  pseudofolder: "arc_ovh",
  container: "arc_ovh_test_container",
  default_tempurl_ttl: (30 * 24 * 60 * 60), # 30 days default time to live for signed urls.
  version_timeout: (60 * 3 * 1000) # 3 minutes


test/arc_ovh_test.exs
defmodule ArcOvhTest do
  use ArcOvh.TestCase, async: :false
  @client ArcOvh.Client.Cloudstorage

  test "Image can be stored as an upload to OVH public cloud storage" do
    DummyDefinition.store(image())
    server_objects = @client.swift().list_objects!(test_container())
    expected_server_object = "#{test_pseudofolder()}/#{Path.basename(image())}"
    expected_object_binary = File.read!(image())
    actual_object_binary = @client.swift().download_file!(expected_server_object, test_container())
    expected_image_size = original_image_size()
    downloaded_file_path = Path.join(tmp_path(), expected_server_object)
    File.touch!(downloaded_file_path) && File.write!(downloaded_file_path, actual_object_binary)
    actual_image_size = Fastimage.size(downloaded_file_path)

    assert(expected_server_object in server_objects)
    assert(expected_object_binary == actual_object_binary)
    assert(expected_image_size == actual_image_size)
  end


  test "Image can be resized and then stored as an upload to OVH public cloud storage" do
    DummyDefinitionResize.store(image())
    server_objects = @client.swift().list_objects!(test_container())
    expected_server_object = "#{test_pseudofolder()}/#{Path.basename(image())}"
    expected_object_binary = File.read!(image())
    actual_object_binary = @client.swift().download_file!(expected_server_object, test_container())
    expected_image_size = original_image_size()
    downloaded_file_path = Path.join(tmp_path(), expected_server_object)
    File.touch!(downloaded_file_path) && File.write!(downloaded_file_path, actual_object_binary)
    actual_image_size = Fastimage.size(downloaded_file_path)

    assert(expected_server_object in server_objects)
    refute(expected_object_binary == actual_object_binary)
    refute(expected_image_size == actual_image_size)
    assert(%{height: 100, width: 100} == actual_image_size)
  end


  test "Url generation for :original and :thumbnail, unsigned" do
    DummyDefinitionThumbnail.store({image(), %{id: 1}})
    public_url = @client.swift().get_public_url()
    actual_original_url = DummyDefinitionThumbnail.url({image(), %{id: 1}}, :original)
    actual_thumbnail_url = DummyDefinitionThumbnail.url({image(), %{id: 1}}, :thumbnail)

    expected_original_url =
    "#{public_url}/#{test_container()}/#{test_pseudofolder()}/" <>
    "#{1}/#{Path.basename(image()) |> Path.rootname()}_original#{Path.extname(image())}"

    expected_thumnbnail_url =
    "#{public_url}/#{test_container()}/#{test_pseudofolder()}/" <>
    "#{1}/#{Path.basename(image()) |> Path.rootname()}_thumbnail#{Path.extname(image())}"

    assert(expected_original_url == actual_original_url)
    assert(expected_thumnbnail_url == actual_thumbnail_url)
  end


  test "Url generation for :original and :thumbnail, signed" do
    DummyDefinitionThumbnail.store({image(), %{id: 1}})
    public_url = @client.swift().get_public_url()
    actual_original_url = DummyDefinitionThumbnail.url({image(), %{id: 1}}, :original, signed: :true)
    actual_thumbnail_url = DummyDefinitionThumbnail.url({image(), %{id: 1}}, :thumbnail, signed: :true)

    expected_original_binary = File.read!(image())
    actual_original_binary = HTTPipe.get!(actual_original_url) |> Map.get(:response) |> Map.get(:body)

    expected_original_url =
    "#{public_url}/#{test_container()}/#{test_pseudofolder()}/" <>
    "#{1}/#{Path.basename(image()) |> Path.rootname()}_original#{Path.extname(image())}"

    expected_thumnbnail_url =
    "#{public_url}/#{test_container()}/#{test_pseudofolder()}/" <>
    "#{1}/#{Path.basename(image()) |> Path.rootname()}_thumbnail#{Path.extname(image())}"

    assert(String.contains?(actual_original_url, expected_original_url))
    assert(String.contains?(actual_thumbnail_url, expected_thumnbnail_url))
    assert(expected_original_binary == actual_original_binary)
  end

  test "Image deletion" do
    {:ok, object_basename} = DummyDefinition.store({image(), %{id: 1}})
    server_object = "#{test_pseudofolder()}/#{object_basename}"
    object_exists_before_deletion? = object_exists?(server_object)

    :ok = DummyDefinition.delete({object_basename, %{id: 1}})
    :timer.sleep(300)

    object_exists_after_deletion? = object_exists?(server_object)
    assert(object_exists_before_deletion? == :true)
    assert(object_exists_after_deletion? == :false)
  end

  test "Url deletion, unsigned" do
    {:ok, object_basename} = DummyDefinitionThumbnail.store({image(), %{id: 1}})
    actual_original_url = DummyDefinitionThumbnail.url({image(), %{id: 1}}, :original)
    actual_thumbnail_url = DummyDefinitionThumbnail.url({image(), %{id: 1}}, :thumbnail)

    original_server_object = actual_original_url |> String.split(test_container() <> "/") |> List.last()
    thumbnail_server_object = actual_thumbnail_url |> String.split(test_container() <> "/") |> List.last()

    object_exists_before_deletion_original? = object_exists?(original_server_object)
    object_exists_before_deletion_thumbnail? = object_exists?(thumbnail_server_object)

    :ok = DummyDefinitionThumbnail.delete({object_basename, %{id: 1}})
    :timer.sleep(300)

    object_exists_after_deletion_original? = object_exists?(original_server_object)
    object_exists_after_deletion_thumbnail? = object_exists?(thumbnail_server_object)

    assert(object_exists_before_deletion_original? == :true)
    assert(object_exists_before_deletion_thumbnail? == :true)
    assert(object_exists_after_deletion_original? == :false)
    assert(object_exists_after_deletion_thumbnail? == :false)
  end

  test "Url deletion, unsigned, without scope" do
    {:ok, _object_basename} = DummyDefinitionThumbnail.store({image(), %{id: 1}})
    actual_original_url = DummyDefinitionThumbnail.url({image(), %{id: 1}}, :original)
    actual_thumbnail_url = DummyDefinitionThumbnail.url({image(), %{id: 1}}, :thumbnail)

    original_server_object = actual_original_url |> String.split(test_container() <> "/") |> List.last()
    thumbnail_server_object = actual_thumbnail_url |> String.split(test_container() <> "/") |> List.last()

    object_exists_before_deletion_original? = object_exists?(original_server_object)
    object_exists_before_deletion_thumbnail? = object_exists?(thumbnail_server_object)

    :ok = DummyDefinitionThumbnail.delete(actual_original_url)
    :ok = DummyDefinitionThumbnail.delete(actual_thumbnail_url)
    :timer.sleep(300)

    object_exists_after_deletion_original? = object_exists?(original_server_object)
    object_exists_after_deletion_thumbnail? = object_exists?(thumbnail_server_object)

    assert(object_exists_before_deletion_original? == :true)
    assert(object_exists_before_deletion_thumbnail? == :true)
    assert(object_exists_after_deletion_original? == :false)
    assert(object_exists_after_deletion_thumbnail? == :false)
  end


  test "Url deletion, signed" do
    {:ok, object_basename} = DummyDefinitionThumbnail.store({image(), %{id: 1}})
    actual_original_url = DummyDefinitionThumbnail.url({image(), %{id: 1}}, :original, signed: :true)
    actual_thumbnail_url = DummyDefinitionThumbnail.url({image(), %{id: 1}}, :thumbnail, signed: :true)

    expected_original_binary = File.read!(image())
    actual_original_binary_before = HTTPipe.get!(actual_original_url) |> Map.get(:response)
    actual_thumbnail_binary_before = HTTPipe.get!(actual_thumbnail_url) |> Map.get(:response)
    thumbnail_size = Fastimage.size(actual_thumbnail_url)

    :ok = DummyDefinitionThumbnail.delete({object_basename, %{id: 1}})
    :timer.sleep(300)

    actual_original_binary_after = HTTPipe.get!(actual_original_url) |> Map.get(:response)
    actual_thumbnail_binary_after = HTTPipe.get!(actual_thumbnail_url) |> Map.get(:response)

    assert(expected_original_binary == Map.get(actual_original_binary_before, :body))
    assert(actual_original_binary_before.status_code == 200)
    assert(actual_thumbnail_binary_before.status_code == 200)
    assert(%{height: 100, width: 100} == thumbnail_size)
    assert(actual_original_binary_after.status_code == 404)
    assert(actual_thumbnail_binary_after.status_code == 404)
  end


  # private

  defp image(), do: Path.join(__DIR__, "./fixtures/bundoran.jpg") |> Path.expand()

  defp tmp_path(), do: Path.join(__DIR__, "./tmp") |> Path.expand()

  defp test_container(), do: Application.get_env(:arc, :container, "arc_ovh_test_container")

  defp test_pseudofolder(), do: Application.get_env(:arc, :pseudofolder, "arc_ovh")

  defp original_image_size(), do: Fastimage.size(image())

  defp object_exists?(object) do
    @client.swift().list_objects!(test_container())
    |> Enum.member?(object)
  end

end

test/support/test_case.ex
defmodule ArcOvh.TestCase do
  @moduledoc :false
  use ExUnit.CaseTemplate, async: :false
  alias Openstex.Swift.V1
  @client ArcOvh.Client.Cloudstorage
  @tmp_path Path.join(__DIR__, "../tmp") |> Path.expand()

  setup() do
    {:ok, containers} = @client.swift().list_containers()
    (containers == []) && create_container!()
    {:ok, containers} = @client.swift().list_containers()
    (not test_container() in containers) && create_container!()
    :ok = Path.join(@tmp_path, test_pseudofolder()) |> File.mkdir_p!()
    on_exit(fn() ->
      delete_container!()
      File.rm_rf!(@tmp_path)
    end)
    :ok
  end

  # private
  defp test_container(), do: Application.get_env(:arc, :container, "arc_ovh_test_container")
  defp test_pseudofolder(), do: Application.get_env(:arc, :pseudofolder, "arc_ovh")

  defp create_container!() do
    V1.create_container(test_container(), @client.swift().get_account())
    |> @client.request!()
  end

  defp delete_container!() do
    if test_container() in @client.swift().list_containers!() do
      @client.swift().list_pseudofolders!(test_container())
      |> Enum.each(fn(pseudofolder) ->
        @client.swift().delete_pseudofolder(pseudofolder, test_container())
      end)
      @client.swift().list_objects!("", test_container())
      |> Enum.each(fn(obj) ->
        @client.swift().delete_object(obj, test_container())
      end)
      V1.delete_container(test_container(), @client.swift().get_account())
      |> @client.request!()
    end
  end

end


mix.exs
defmodule ArcOvh.Mixfile do
  use Mix.Project
  @elixir_versions "~> 1.5"
  @version "0.2.0"

  def project do
    [
      app: :arc_ovh,
      version: @version,
      source_url: "https://github.com/stephenmoloney/arc_ovh",
      elixir:  @elixir_versions,
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      description: description(),
      elixirc_paths: elixirc_paths(Mix.env),
      package: package(),
      deps: deps(),
      docs: docs()
     ]
  end

  def application do
    [
      mod: {ArcOvh.Application, []},
      extra_applications: []
    ]
  end


  defp deps do
    [
      {:arc, "~> 0.8"},
      {:openstex_adapters_ovh, "~> 0.4"},
      {:mime, "~> 1.1"},

       # dev/test deps
      {:markdown, github: "devinus/markdown", only: :dev},
      {:ex_doc,  "~> 0.18", only: :dev},
      {:fastimage, "~> 0.0.7", only: :test}
    ]
  end


  defp description() do
    ~s"""
    An ovh storage adapter for arc.
    """
  end

  defp docs do
    [
     extras: ["README.md"]
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_),     do: ["lib"]


  defp package() do
    %{
      licenses: ["MIT"],
      maintainers: ["Stephen Moloney"],
      links: %{ "GitHub" => "https://github.com/stephenmoloney/arc_ovh"},
      files: ~w(lib mix.exs README* LICENSE* CHANGELOG*)
     }
  end


end


config/dev.exs
use Mix.Config

config :arc_ovh, ArcOvh.Client.Cloudstorage,
    adapter: Openstex.Adapters.Ovh,
    ovh: [
      application_key: System.get_env("APPLICATION_KEY"),
      application_secret: System.get_env("APPLICATION_SECRET"),
      consumer_key: System.get_env("CONSUMER_KEY")
    ],
    keystone: [
      tenant_id: System.get_env("TENANT_ID"), # mandatory, corresponds to an ovh project id or ovh servicename
      user_id: System.get_env("USER_ID"), # optional, if absent a user will be created using the ovh api.
      endpoint: "https://auth.cloud.ovh.net/v2.0"
    ],
    swift: [
      account_temp_url_key1: System.get_env("TEMP_URL_KEY1"), # defaults to :nil if absent
      account_temp_url_key2: System.get_env("TEMP_URL_KEY2"), # defaults to :nil if absent
      region: :nil
    ],
    hackney: [
      timeout: 20000,
      recv_timeout: 40000
    ]

config :httpipe,
  adapter: HTTPipe.Adapters.Hackney

config :arc,
  storage: Arc.Storage.Ovh.Cloudstorage,
  client: ArcOvh.Client.Cloudstorage,
  container: "dev_container",
  default_tempurl_ttl: (30 * 24 * 60 * 60), # 30 days default time to live for signed urls.
  version_timeout: (60 * 3 * 1000) # 3 minutes
